### PR TITLE
Add support for editing scheme command line arguments

### PR DESCRIPTION
**Pull request by rahul-malik**
This PR adds support for readings/writing command line arguments in scheme actions when available (Launch, Test and Profile).

Resolves https://github.com/xcodeswift/xcproj/issues/165

### Short description 📝
Depending on the project, you may wish to have command line arguments passed to scheme actions to affect their runtime behavior. This patch adds command line options to relevant

### Solution 📦
Add a new class to represent command line arguments in `XCScheme.swift`. Update the relevant actions to read / write the proper XML data

### Implementation 👩‍💻👨‍💻
- [x] Create the `CommandLineArguments` class to represent the XML element in `xcscheme` files
- [x] Update `LaunchAction`, `TestAction` and `ProfileAction` to have an optional command line arguments property (`CommandLineArguments?`)
- [x] Update `init` and `xmlElement` methods to read / write `CommandLineArguments`
- [x] Add use-case to `XCScheme` test fixture 

### GIF
gif

---
This change is [Reviewable](https://reviewable.io/reviews/xcodeswift/xcproj/167)

### CODE REVIEW

File: Sources/xcproj/XCScheme.swift [hunk 1]
```diff
@@ -346,6 +398,9 @@ final public class XCScheme {
             self.useCustomWorkingDirectory = element.attributes["useCustomWorkingDirectory"] == "YES"
             self.debugDocumentVersioning = element.attributes["debugDocumentVersioning"] == "YES"
             self.buildableProductRunnable = try BuildableProductRunnable(element: element["BuildableProductRunnable"])
+            if let commandlineOptions = element.children.first(where: { $0.name == "CommandLineArguments" }) {
```
Comment (yonaskolb) [inline Sources/xcproj/XCScheme.swift]: I think AEXMLElement has some query subscripts doesn't it? (I haven't used it in a while)
Something like `if let commandLineOptions = element["CommandLineArguments"]` (no idea if this compiles)

File: Sources/xcproj/XCScheme.swift [hunk 2]
```diff
@@ -122,6 +122,42 @@ final public class XCScheme {
...
+
+        public init(arguments args:[CommandLineArgument]) {
+            self.arguments = args
+        }
+
+        init(element: AEXMLElement) {
+            self.arguments = element.children.map { elt in
+                guard let argName = elt.attributes["argument"],
+                      let argEnabledRaw = elt.attributes["isEnabled"] else {
+                    fatalError("Invalid Commandline Argument Element")
```
Comment (pepibumur) [inline Sources/xcproj/XCScheme.swift]: I think we should handle this throwing an error that could potentially be handled from upper levels.

File: Sources/xcproj/XCScheme.swift [hunk 3]
```diff
@@ -356,6 +411,9 @@ final public class XCScheme {
                                                     "useCustomWorkingDirectory": useCustomWorkingDirectory.xmlString,
                                                     "debugDocumentVersioning": debugDocumentVersioning.xmlString])
             element.addChild(buildableProductRunnable.xmlElement())
+            if let cliArguments = commandlineArguments {
```
Comment (pepibumur) [inline Sources/xcproj/XCScheme.swift]: I think we can use the same name:
```swift
if let commandlineArguments = commandlineArguments {

}
```

File: Sources/xcproj/XCScheme.swift [hunk 4]
```diff
@@ -398,6 +460,10 @@ final public class XCScheme {
                 .all?
                 .map(TestableReference.init) ?? []
             self.macroExpansion = try BuildableReference(element: element["MacroExpansion"]["BuildableReference"])
+
+            if let commandlineOptions = element.children.first(where: { $0.name == "CommandLineArguments" }) {
```
Comment (pepibumur) [inline Sources/xcproj/XCScheme.swift]: If @yonaskolb is right regarding the XML API to find the children don't forget to update this one as well.

File: Tests/xcprojTests/XCSchemeSpec.swift [hunk 5]
```diff
@@ -60,12 +60,21 @@ final class XCSchemeIntegrationSpec: XCTestCase {
         XCTAssertEqual(scheme.testAction?.macroExpansion.blueprintName, "iOS")
         XCTAssertEqual(scheme.testAction?.macroExpansion.referencedContainer, "container:Project.xcodeproj")
         XCTAssertEqual(scheme.testAction?.macroExpansion.referencedContainer, "container:Project.xcodeproj")
+        XCTAssertNotNil(scheme.testAction?.commandlineArguments)
+        if let testCLIArgs = scheme.testAction?.commandlineArguments {
```
Comment (pepibumur) [inline Tests/xcprojTests/XCSchemeSpec.swift]: @rahul-malik I think we can simplify this by adding a custom assertion:

```swift
let commandLineArguments = try XCTAssertNotNilAndUnwrap(scheme.testAction?.commandlineArguments)
XCTAssertTrue(commandLineArguments.count > 0)
XCTAssertEqual(commandLineArguments[0].name, "MyTestArgument")
XCTAssertTrue(commandLineArguments[0].enabled)
```
By doing that we avoid the testing in our tests. What do you think?

Comment (pepibumur) [reply Tests/xcprojTests/XCSchemeSpec.swift]: Same comment applies to the other tests in this file.

**Comment by pepibumur** [comment]
@rahul-malik just wondering, do you have access to write to this repository? You should have it, if not let me know. It'll be faster if you can just push to the repository.

**Comment by rahul-malik** [comment]
@pepibumur - I'll create branches from the main repo on future PRs. I already had my fork locally so I just wrote this patch there. 

All comments should be addressed, let me know if anything else should be updated

**Comment by pepibumur** [comment]
Reviewed 3 of 3 files at r1.
Review status: 1 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[Tests/xcprojTests/XCTestCase+Assertions.swift, line 1 at r3](https://reviewable.io:443/reviews/xcodeswift/xcproj/167#-KzosZhMgkGyzORMk2T-:-KzosZhMgkGyzORMk2T0:bqd6pl6) ([raw file](https://github.com/xcodeswift/xcproj/blob/7d2d9571eca43be63f3ba9741da4c5ce45b36cdc/Tests/xcprojTests/XCTestCase+Assertions.swift#L1)):*
> ```Swift
> //
> ```

Don't forget to remove the comments in the header :stuck_out_tongue:  @rahul-malik 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xcodeswift/xcproj/167)*

**Pull request merged**